mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java

```diff
@@ -334,94 +332,102 @@ private boolean userHasLookupRightsOn(Mailbox mailbox, MailboxSession session) {
 
     @Override
     public Optional<MailboxId> createMailbox(MailboxPath mailboxPath, MailboxSession mailboxSession) throws MailboxException {
+        return MailboxReactorUtils.blockOptional(createMailboxReactive(mailboxPath, mailboxSession));
+    }
+
+    public Mono<MailboxId> createMailboxReactive(MailboxPath mailboxPath, MailboxSession mailboxSession) {
         LOGGER.debug("createMailbox {}", mailboxPath);
 
-        assertMailboxPathBelongToUser(mailboxSession, mailboxPath);
+        return assertMailboxPathBelongToUserReactive(mailboxSession, mailboxPath)
+            .then(doCreateMailboxReactive(mailboxPath, mailboxSession));
+    }
 
+    private Mono<MailboxId> doCreateMailboxReactive(MailboxPath mailboxPath, MailboxSession mailboxSession) {
         if (mailboxPath.getName().isEmpty()) {
             LOGGER.warn("Ignoring mailbox with empty name");
+            return Mono.empty();
         } else {
-            MailboxPath sanitizedMailboxPath = mailboxPath.sanitize(mailboxSession.getPathDelimiter());
-            sanitizedMailboxPath.assertAcceptable(mailboxSession.getPathDelimiter());
-
-            if (block(mailboxExists(sanitizedMailboxPath, mailboxSession))) {
-                throw new MailboxExistsException(sanitizedMailboxPath.asString());
-            }
-
-            List<MailboxId> mailboxIds = createMailboxesForPath(mailboxSession, sanitizedMailboxPath);
-
-            if (!mailboxIds.isEmpty()) {
-                return Optional.ofNullable(Iterables.getLast(mailboxIds));
+            try {
+                MailboxPath sanitizedMailboxPath = mailboxPath.sanitize(mailboxSession.getPathDelimiter());
+                sanitizedMailboxPath.assertAcceptable(mailboxSession.getPathDelimiter());
+
+                return mailboxExists(sanitizedMailboxPath, mailboxSession)
+                    .flatMap(exists -> {
+                        if (exists) {
+                            return Mono.error(new MailboxExistsException(sanitizedMailboxPath.asString()));
+                        } else {
+                            return createMailboxesForPath(mailboxSession, sanitizedMailboxPath).last();
+                        }
+                    });
+            } catch (MailboxNameException e) {
+                return Mono.error(e);
             }
```

Comment:
```suggestion
            return Mono.fromCallable(() -> mailboxPath.sanitize(mailboxSession.getPathDelimiter()))
                .map(Throwing.function(sanitizedMailboxPath -> sanitizedMailboxPath.assertAcceptable(mailboxSession.getPathDelimiter())))
                .flatMap(sanitizedMailboxPath -> mailboxExists(sanitizedMailboxPath, mailboxSession)
                    .flatMap(exists -> {
                        if (exists) {
                            return Mono.error(new MailboxExistsException(sanitizedMailboxPath.asString()));
                        } else {
                            return createMailboxesForPath(mailboxSession, sanitizedMailboxPath).last();
                        }
                    }))
                .onErrorResume(MailboxNameException.class, Mono::error);
```

Reply:
No I want to avoid an uneeded flatmap bringing asynchronous semantic in asynchronous code.